### ISSUE TITLE
Revisit `handle_string_response` implementation. The current one is very confusing.

### ISSUE THREAD

**Issue by ktonga** [ensime/ensime-vim#330]
I think the handling around text responses is currently hard to follow, it handles very different kind of messages and the code branching is base on non-clearly-named global state.

**Comment by ches** [comment]
A saner approach than `call_options` and the way class attributes are sometimes set as flags in some far-off location in the code (which is also dependent on sequential processing of the queue) probably boils down to a fundamental design change that I'm not sure I personally will even try to rework before working instead on my side project to rewrite the client core—I'm attempting to use an event loop instead of a brute-polling thread, seems like the efficient choice for something I/O-bound.

`handle_string_response` can at least dispatch to separate functions with appropriate names for handling the different requests that it determines with the state variables, in the shorter term.

Maybe there could be some kind of small interim improvement by maintaining a shared dictionary of `call_id` -> `typehint`, so an overloaded response message type like `StringResponse` could look up more precisely what the request type was instead those icky flag attributes?

**Comment by ktonga** [comment]
Yes, that would be enough for a short term solution. And maybe keep a dict `call_id` -> `buffer name` for source formatting to always replace the content in the right buffer.

```python
self.editor.replaceText(buffer_names[call_id], payload['text'])
buffer_names.pop(call_id, None)
```

**Comment by ches** [comment]
Ah that's what you meant about changing buffer, I had thought you meant editing the current buffer's content. That's a good point.

So mechanically this stuff is basically what `call_options` already is. We could piggyback on that, possibly rename it to something that reflects the expanded significance.

**Comment by ktonga** [comment]
Oh I wasn't aware of `call_options` it's so global that it is almost invisible. A good approach to give it more visibility would be to add it to the handler's params.

```python
    def handle_incoming_response(self, call_id, payload):
        ...
        # make it safer (handle missing entries)
        options = self.call_options[call_id]
        handler(call_id, payload, options)
        ....

    # then in some handler implementation
    def handle_symbol_info(self, call_id, payload, options):
        ...
        display = options.get("display")
        ...

```
